File: Android/melody-runtime/src/main/java/com/melody/runtime/components/MelodyToggle.kt
package com.melody.runtime.components

import androidx.compose.material3.Switch
import androidx.compose.material3.SwitchDefaults
import androidx.compose.material3.Text
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.width
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.melody.core.schema.ComponentDefinition
import com.melody.core.schema.resolved
import com.melody.runtime.engine.LuaValue
import com.melody.runtime.renderer.LocalIsDisabled
import com.melody.runtime.renderer.LocalScreenState
import com.melody.runtime.renderer.LocalThemeColors

@Composable
fun MelodyToggle(definition: ComponentDefinition, onChanged: (() -> Unit)? = null) {
    val screenState = LocalScreenState.current
    val isDisabled = LocalIsDisabled.current
    val themeColors = LocalThemeColors.current
    val style = definition.style
    val stateKey = definition.stateKey
    val label = definition.label.resolved ?: ""
    val tintColor = style?.color.resolved?.let { parseColor(it, themeColors) }

    val isOn = if (stateKey != null) {
        screenState.slot(stateKey).value.boolValue ?: false
    } else {
        false
    }

    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.SpaceBetween,
        modifier = Modifier.melodyStyle(style)
    ) {
        if (label.isNotEmpty()) {
            Text(label)
            Spacer(modifier = Modifier.width(8.dp))
        }
        Switch(
            checked = isOn,
            enabled = !isDisabled,
            onCheckedChange = { newValue ->
                if (stateKey != null) {
                    screenState.set(stateKey, LuaValue.BoolVal(newValue))
                }
                onChanged?.invoke()
            },
            colors = if (tintColor != null) {
                SwitchDefaults.colors(checkedTrackColor = tintColor)
            } else {
                SwitchDefaults.colors()
            }
        )
    }
}

File: Sources/Runtime/Components/MelodyToggle.swift
import SwiftUI
import Core

/// Renders a toggle switch bound to a screen state key.
struct MelodyToggle: View {
    let definition: ComponentDefinition
    var onChanged: (() -> Void)? = nil

    @Environment(\.screenState) private var screenState
    @Environment(\.themeColors) private var themeColors

    var body: some View {
        Toggle(resolvedLabel, isOn: binding)
            .tint(tintColor)
            .onChange(of: binding.wrappedValue) { _, _ in
                onChanged?()
            }
    }

    private var resolvedLabel: String {
        definition.label.resolved ?? ""
    }

    private var binding: Binding<Bool> {
        guard let key = definition.stateKey else {
            return .constant(false)
        }
        return Binding(
            get: { screenState.get(key: key).boolValue ?? false },
            set: { screenState.set(key: key, value: .bool($0)) }
        )
    }

    private var tintColor: Color? {
        guard let value = definition.style?.color.resolved else { return nil }
        return Color(hex: StyleResolver.colorHex(value, themeColors: themeColors))
    }
}
